### PR TITLE
Teach RemoteMirror how to project enum values

### PR DESCRIPTION
**Pull request by tbkka**
This adds two new functions to the SwiftRemoteMirror
facility that support inspecting enum values.

Currently, these support non-payload enums and
single-payload enums, including nested enums and
payloads with struct, tuple, and reference payloads.
In particular, it handles nested `Optional` types.

TODO: Multi-payload enums use different strategies for
encoding the cases that aren't yet supported by this
code.

Resolves rdar://59961527

```
/// Projects the value of an enum.
///
/// Takes the address and typeref for an enum and determines the
/// index of the currently-selected case within the enum.
///
/// Returns true iff the enum case could be successfully determined.
/// In particular, note that this code may fail for valid in-memory data
/// if the compiler is using a strategy we do not yet understand.
SWIFT_REMOTE_MIRROR_LINKAGE
int swift_reflection_projectEnumValue(SwiftReflectionContextRef ContextRef,
                                      swift_addr_t EnumAddress,
                                      swift_typeref_t EnumTypeRef,
                                      int *CaseIndex);

/// Finds information about a particular enum case.
///
/// Given an enum typeref and index of a case, returns:
/// * Typeref of the associated payload or zero if there is no payload
/// * Name of the case if known.
///
/// The Name points to a freshly-allocated C string on the heap.  You
/// are responsible for freeing the string (via `free()`) when you are finished.
SWIFT_REMOTE_MIRROR_LINKAGE
int swift_reflection_getEnumCaseTypeRef(SwiftReflectionContextRef ContextRef,
                                        swift_typeref_t EnumTypeRef,
                                        int CaseIndex,
                                        char **CaseName,
                                        swift_typeref_t *PayloadTypeRef);
```

### CODE REVIEW

**Comment by tbkka** [comment]
This is not quite ready to merge:  It currently hard-codes some pointer properties that need to be queried from the target.

File: include/swift/Remote/InProcessMemoryReader.h [hunk 1]
```diff
@@ -30,18 +30,40 @@ class InProcessMemoryReader final : public MemoryReader {
...
+    }
+    case DLQ_GetSizeSize: {
+      auto result = static_cast<uint8_t *>(outBuffer);
+      *result = sizeof(size_t);
+      return true;
+    }
+    case DLQ_GetObjCReservedLowBits: {
+      auto result = static_cast<uint8_t *>(outBuffer);
+#if __APPLE__ && __x86_64__
+      // ObjC on 64-bit macOS reserves the bottom bit of all pointers.
```
Comment (dcci) [inline include/swift/Remote/InProcessMemoryReader.h]: What about iOS?

Comment (tbkka) [reply include/swift/Remote/InProcessMemoryReader.h]: As far as I understand, iOS Obj-C reserves high bits, not low bits.  So this literally only applies to 64-bit macOS, no other platforms.

Comment (dcci) [reply include/swift/Remote/InProcessMemoryReader.h]: Thanks for clarifying.

File: include/swift/Reflection/TypeLowering.h [hunk 2]
```diff
@@ -155,6 +173,17 @@ class BuiltinTypeInfo : public TypeInfo {
...
 
+  bool readExtraInhabitantIndex(remote::MemoryReader &reader,
+                                remote::RemoteAddress address,
+                                int *extraInhabitantIndex) const {
+    if (getNumExtraInhabitants() == 0) {
+      *extraInhabitantIndex = -1;
+      return true;
+    }
+    // XXX Has extra inhabitants, so it must be a pointer?
+    return reader.readHeapObjectExtraInhabitantIndex(address, extraInhabitantIndex);
```
Comment (tbkka) [inline include/swift/Reflection/TypeLowering.h]: I'm not sure whether a `BuiltinTypeInfo` with extra inhabitants can ever be a function pointer or whether it's always just a regular pointer to heap data.  (There are different rules for managing XIs for function pointers.)

File: include/swift/Reflection/TypeLowering.h [hunk 3]
```diff
@@ -206,6 +239,16 @@ class ReferenceTypeInfo : public TypeInfo {
...
   }
 
+  bool readExtraInhabitantIndex(remote::MemoryReader &reader,
+                                remote::RemoteAddress address,
+                                int *extraInhabitantIndex) const {
+    if (getNumExtraInhabitants() == 0) {
+      *extraInhabitantIndex = -1;
+      return true;
+    }
+    return reader.readHeapObjectExtraInhabitantIndex(address, extraInhabitantIndex);
```
Comment (tbkka) [inline include/swift/Reflection/TypeLowering.h]: I suspect `ReferenceTypeInfo` is used for both function pointers and heap pointers, but I'm not quite sure how to distinguish those cases.

**Comment by tbkka** [comment]
I think I've worked out the pointer properties as well as I can for now.  I'm still checking some details around XI handling for different kinds of pointers, but otherwise I think this is mergeable (pending CI, of course).

**Comment by tbkka** [comment]
@swift-ci Please test macOS

**Comment by tbkka** [comment]
@swift-ci Please test Linux

**Comment by tbkka** [comment]
I think this is ready to merge.  The tests could be a little more comprehensive, but the API seems to be working correctly for the current tests on both macOS x86_64 and the 32-bit iOS simulator.

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
@swift-ci Please test macOS

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
@swift-ci Please test macOS

**Comment by tbkka** [comment]
@swift-ci Please test macOS

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by tbkka** [comment]
@swift-ci Please smoke test Linux

**Comment by rintaro** [comment]
This caused test failure in "Swift in the OS" bot
https://ci.swift.org/job/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/1589/console
@tbkka Could you take a look? rdar://problem/60183719

**Comment by tbkka** [comment]
PR #30281 should fix this.

File: include/swift/Reflection/ReflectionContext.h [hunk 4]
```diff
@@ -710,6 +711,128 @@ class ReflectionContext
...
+    }
+
+    default:
+      // Unknown record kind.
+      break;
+    }
+    return false;
+  }
+
+  bool getEnumCaseTypeRef(const TypeRef *EnumTR,
```
Comment (slavapestov) [inline include/swift/Reflection/ReflectionContext.h]: I think this API is redundant. In the C API, we already have swift_reflection_childOfTypeRef()

**Pull request merged**